### PR TITLE
[protocol] check address format

### PR DESCRIPTION
**Pull request by dustinxie**
# Description
as title

## Type of change
- [x] Bug fix (non-breaking change which fixes an issue)

# How Has This Been Tested?
- [x] make test

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: action/protocol/account/transfer.go [hunk 1]
```diff
@@ -63,19 +65,29 @@ func (p *Protocol) handleTransfer(ctx context.Context, act action.Action, sm pro
...
+	}
 	if err != nil {
 		return nil, errors.Wrapf(err, "failed to decode recipient address %s", tsf.Recipient())
 	}
 	recipientAcct, err := accountutil.LoadAccount(sm, recipientAddr)
+	if !fCtx.CheckLegacyAddress {
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to load address %s", tsf.Recipient())
+		}
+	}
```
Comment (dustinxie) [inline action/protocol/account/transfer.go:82]: this is a long-time unnoticed issue, in the case
```
err != nil && recipient is NOT a contract
```
the err escaped and will continue from line 109

after the HF, we can remove the `if !fCtx.CheckLegacyAddress` and always apply the check

**Comment by Liuhaai** [comment]
has addr in staking action similar issue?

File: action/protocol/account/transfer.go (hunk 1 above)
Comment (Liuhaai) [reply action/protocol/account/transfer.go:82]: is it possible remove the height checking after the hardfork?

**Review by CoderZhi — COMMENTED** [review]
https://github.com/iotexproject/iotex-core/blob/master/action/consignment_transfer.go#L121
https://github.com/iotexproject/iotex-core/blob/master/action/transfer.go#L153
actions will be rejected if it contains any legacy format.

it is not an easy fix, please go through all places again, https://github.com/iotexproject/iotex-core/search?q=address.FromString

File: action/execution_test.go [hunk 2]
```diff
@@ -62,20 +62,6 @@ func TestExecutionSanityCheck(t *testing.T) {
...
-			uint64(1),
-			big.NewInt(0),
-			uint64(0),
-			big.NewInt(0),
-			[]byte{},
-		)
-		require.NoError(err)
-		require.Contains(ex.SanityCheck().Error(), "error when validating contract's address")
-	})
-
```
Comment (dustinxie) [inline action/execution_test.go:78]: added test in `generic_validator_test.go`

File: action/action.go [hunk 3]
```diff
@@ -34,7 +34,7 @@ type (
 	}
 
 	hasDestination interface {
-		Destination() string
+		Destination() (string, bool)
```
Comment (CoderZhi) [inline action/action.go:37]: what's the point of the new return value?

File: action/execution.go [hunk 4]
```diff
@@ -68,7 +68,7 @@ func NewExecution(
 func (ex *Execution) Contract() string { return ex.contract }
 
 // Destination returns a contract address
-func (ex *Execution) Destination() string { return ex.contract }
+func (ex *Execution) Destination() (string, bool) { return ex.contract, ex.contract != EmptyAddress }
```
Comment (Liuhaai) [inline action/execution.go:71]: What are the benefits if the interface is changed?

**Comment by Liuhaai** [comment]
The err handle could be in https://github.com/iotexproject/iotex-core/blob/master/state/factory/workingset.go#L457 instead of in  the actpool?

File: blockindex/indexer.go [hunk 5]
```diff
@@ -393,7 +400,7 @@ func (x *blockIndexer) getIndexerForAddr(addr []byte, batch bool) (db.CountingIn
 }
 
 // indexAction builds index for an action
-func (x *blockIndexer) indexAction(actHash hash.Hash256, elp action.SealedEnvelope, insert bool) error {
+func (x *blockIndexer) indexAction(actHash hash.Hash256, elp action.SealedEnvelope, insert, tolerateLegacyAddress bool) error {
```
[the remark this answers is not in the document]
Comment (dustinxie) [reply blockindex/indexer.go:403]: need more code/work outside and logic will be complicated, would rather keep it simple, `insert` is also a `bool`

**Review by Liuhaai — APPROVED** [review]

File: config/config.go [hunk 6]
```diff
@@ -112,7 +112,7 @@ var (
 			MaxNumActsPerPool:  32000,
 			MaxGasLimitPerPool: 320000000,
 			MaxNumActsPerAcct:  2000,
-			ActionExpiry:       10 * time.Minute,
+			ActionExpiry:       time.Minute,
```
Comment (dustinxie) [inline config/config.go:115]: with this PR, a transfer/exec with bad address won't pass validation, but will remain in actpool

thinking to reduce the action expiry time, to speed up removing the pending bad tx out of actpool

Comment (dustinxie) [reply config/config.go:115]: 1 min already translates to 12 chances of attempt (at 5 second a block)

if a tx failed for 1 min, it's meaningless to leave it 10 mins

File: config/config.go [hunk 7]
```diff
@@ -112,7 +112,7 @@ var (
 			MaxNumActsPerPool:  32000,
 			MaxGasLimitPerPool: 320000000,
 			MaxNumActsPerAcct:  2000,
-			ActionExpiry:       10 * time.Minute,
+			ActionExpiry:       3 * time.Minute,
```
Comment (dustinxie) [inline config/config.go:115]: with this PR, transfer/exec with bad address will remain in actpool

thinking to reduce the expire time, to speed up removing it out of actpool

Comment (Liuhaai) [reply config/config.go:115]: Don't integrate this change in this pr

Comment (dustinxie) [reply config/config.go:115]: yes, we can discuss this, and another PR if needed

File: action/protocol/execution/protocol.go [hunk 8]
```diff
@@ -85,7 +85,20 @@ func (p *Protocol) Validate(_ context.Context, act action.Action, _ protocol.Sta
 	if exec.TotalSize() > ExecutionSizeLimit {
 		return action.ErrOversizedData
 	}
-	return nil
+	var (
+		dest = exec.Contract()
+		fCtx = protocol.MustGetFeatureCtx(ctx)
```
Comment (CoderZhi) [inline action/protocol/execution/protocol.go:90]: `fCtx` is not needed if `dest == action.EmptyAddress`

File: config/config.go (hunk 7 above)
Comment (CoderZhi) [inline config/config.go:115]: revert

**Review by CoderZhi — APPROVED** [review]

**Pull request merged**